python/cudf/cudf/core/frame.py

```diff
@@ -3301,6 +3301,25 @@ def _encode(self):
         keys = self.__class__._from_table(keys)
         return keys, indices
 
+    def _reindex_frame_by_columns(
+        self, existing_column_names, index, num_rows, dtypes=None, deep=False
```

Comment:
Aren't `num_rows` and `index` always `len(self)` and `self.index`?

Reply:
We can create a dataframe by passing in empty `ColumnAccessor`/`data` but a non-empty `index`. In which case following condition takes care of such scenarios: `num_rows=len(index) if data._data.nrows == 0 else data._data.nrows`

Reply:
If not, can we narrow the scope of this function so that it doesn't have to worry about the `index` or `num_rows`? From my perspective, `_reindex_by_columns` should ideally have only the following responsibilities:

* If `columns` is a subset of `self.columns`, return a frame composed of that subset
* If `columns` is a superset, return a frame with extra null columns 

Handling `index` seems to be orthogonal to `_reindex_by_columns` where maybe that should be done in a `_reindex_by_index` function instead.

Reply:
Discussed offline with @shwina and decided to implement `_reindex_frame`